### DISCUSSION TITLE
Issue with Angular v16 - Uncaught ReferenceError: _asyncToGenerator is not defined

### DISCUSSION THREAD

**Discussion by adevinwild** [maplibre/maplibre-gl-js#3210] [Q&A] [answered]
**maplibre-gl-js version**:  `"maplibre-gl": "^3.4.1"`

**browser**: `Chrome`

### Steps to Trigger Behavior

 1. Create an Angular Map component that imports `Map` from `maplibre-gl`
 2. Create a new instance of the `Map` in the `ngAfterViewInit`
 3. Map shows without styles and error pops in the console  : 

### Expected Behavior
It should work ?
### Actual Behavior
Triggers this error 
 
![image](https://github.com/maplibre/maplibre-gl-js/assets/68734155/275ff020-23cb-4ac2-bad9-6eea251d919f)

**Comment by HarelM** [comment]
Please provide a stackblitz reproduction.
Also I advise to use the angular wrapper - ngx-maplibre-gl.
I suspect the problem is with your code as I maintain the ngx wrapper and I know for a fact it is working great.

**Comment by adevinwild** [comment]
Everything in this [example](https://stackblitz.com/edit/angular-1ogzdf?file=src%2Fmain.ts,angular.json) works fine on Stackblitz, but when I download the project and run it locally I get the asyncToGenerator errors , I'm new to Angular
Do I need to add something before running this project ? 
I saw the concept of polyfills too, can it help me fix this error ?

I also cloned your wrapper with Angular, and tried to run it on my computer
No `_asyncToGenerator` error, but I get another error where `getImageData` function is not defined...

**Comment by HarelM** [comment]
Sorry, you'll need to figure this out on your own...

**Comment by neodescis** [comment]
@HarelM , I am seeing this myself. I haven't tried 3.4.1, but it's happening with 3.5.0. I'm investigating now. The error appears to be related to Babel, and I'm guessing also related to the change in 3.4.1 that's using async/await.

**Comment by neodescis** [comment]
I'm going to look into it, but @msbarry , any ideas?

**Reply by msbarry** [reply]
Your target is `es2015` so it's transpiling async/await to other javascript that uses a library that's not being added. I'd suggest either 1) target es2017 or later, we discussed browser support quite a bit in #3185 and found that you most likely don't need to transpile it at all since only a tiny portion of browsers that support webgl don't support async/await
2) or change your tsconfig so it's actually injecting the library it depends on

**Comment by neodescis** [comment]
Well, in my tsconfig the target is actually ES2022, so I'm not sure why it would be transpiling it, but I'll dig a bit and see what I can find.

**Comment by neodescis** [comment]
I did some digging and it seems that Angular is forcing the transpilation to occur because zone.js is not compatible with native async/await:

https://github.com/angular/angular-cli/issues/25708
https://github.com/angular/angular/issues/31730
https://github.com/angular/angular-cli/pull/21552

What I haven't been able to figure out is why it's transpiling but not including the function definition when it does so.

So, @HarelM , have you indeed been able to get 3.4.1 or 3.5.0 to work in Angular yet? Per the above, it seems like it shouldn't work. This appears to be a legitimate issue.

**Comment by neodescis** [comment]
It appears that using the "experimental" Angular browser-esbuild fixes the issue, but I am not able to use it myself because of another unrelated open issue :-(

**Comment by neodescis** [comment]
It seems that the _asyncToGenerator function is actually included, but it's not available to the (transpiled) code on the worker thread because of how it's blob-ified, so it bombs.

**Comment by neodescis** [comment]
Stackblitz must be building things differently, because it's not transpiling the async functions in the example provided by @adevinwild. When I download his project and get it to build (I had to upgrade TypeScript and add "allowSyntheticDefaultImports": true, providing further evidence that Stackblitz must be doing its own thing), I see the same error when I run it.

**Comment by HarelM** [comment]
My mistake, sorry, it seems like the latest changes with the async code in version 3.5 creates problems with angular.
I'm still looking into how this can be solved...
As a workaround, I would hold with upgrading to latest version...

**Comment by msbarry** [comment]
Seems like when you're using zone.js (or any build that transpires maplibre further than it already is) you should probably use the csp builds so that it injects polyfills into the worker properly. Or provide some way to inject polyfills into the worker manually?

**Comment by HarelM** [comment]
Yes, zone seems like the offender here. CSP is a good direction to try, maybe I can have the ngx wrapper code use the csp build somehow...

**Comment by neodescis** [comment]
I've tried to provide a worker through an Angular library before and it wasn't even possible. The library packager was not able to build the worker. Only an "application" was able to do so. This would mean that the consuming application would need to build the worker and provide it to MapLibre somehow, which would make it quite onerous to use, IMO. I don't know if that has changed recently, but I would guess not.

**Comment by neodescis** [comment]
Also, I don't think this translation even needs to occur since zone.js shouldn't need to care about the worker thread, but I don't see any way to turn it off.

**Comment by msbarry** [comment]
I think I'm hitting the same error in create-react-app, it just manifests as `ReferenceError: _ is not defined`

**Reply by HarelM** [reply]
Cc: @Pessimistress let me know if you are seeing this with the react wrapper using the latest maplibre-gl release.

**Reply by Pessimistress** [reply]
Might be similar to [this issue](https://github.com/visgl/react-map-gl/issues/1266)?

Mapbox official docs: https://docs.mapbox.com/mapbox-gl-js/guides/install/#transpiling

**Reply by HarelM** [reply]
I'll check the provided solutions there, it seems similar to what I'm looking for... Thanks @Pessimistress!!

**Reply by neodescis** [reply]
Good suggestion, but I believe Angular is forcing this particular translation despite what's decided via browserlist. Or at least I was unable to bypass it, and the cli babel loader code seems to force it no matter what. I'd be happy to be proven wrong though!

**Reply by HarelM** [reply]
I was not able to solve the problem with any of the suggestion solution there, neither the `!` nor browserlist.
Currently the only solution is to use the CSP version, but that's not a very good solution from a developer perspective... :-(

**Comment by HarelM** [comment]
If I use a local version of the csp worker, I can bypass the problem as the local version of the csp worker gets loaded as is, without any transpilation (it gets loaded on the fly).
This is a very poor solution though as you need to include the worker js code in your angular.json file and probably other weird things...
Maybe there's a way to exclude certain files from transpilation/babel processing somehow... I need to investigate more...

**Comment by sommmen** [comment]
Hello, i'm having the same issue, but i'm currently on Angular V15.
The latest version i can work with is 3.3.1, @maplibre/ngx-maplibre-gl is at version 15.

**Comment by birkskyum** [comment]
For some, dropping zone.js might also be a path forward. zone.js hasn't had an update in 4 years, and manually adding change detection does increase performance. https://github.com/krausest/js-framework-benchmark/pull/1403#issuecomment-1762062968

**Reply by HarelM** [reply]
Removing zone.js is a huge surgery, chances of survival are low for big applications. it's a lot easier to use the csp solution...
Never the less, it's an option...

**Comment by birkskyum** [comment]
- https://github.com/angular/angular-cli/issues/22191

**Comment by msbarry** [comment]
So I guess it's harder to get developers to update old tooling than it is to get end-users to update old browsers 🙃 

Should we have similar guidance that [mapbox gives](https://docs.mapbox.com/mapbox-gl-js/guides/install/#transpiling), for example does `import maplibregl from '!maplibre-gl';` work?

**Reply by HarelM** [reply]
I did not find any of the solutions provided by mapbox to work for our case for Angular, except the CSP, which I find cumbersome...

**Reply by msbarry** [reply]
2 ways I got this to work in a create-react-app app:

1) upgrade from cra v4 to cra v5 which drops this transform
2) the `import '!maplibre-gl'` trick to get webpack to not transpiling the import:

```diff
-import maplibre from 'maplibre-gl';
+// eslint-disable-next-line import/no-webpack-loader-syntax
+import maplibre from '!maplibre-gl';
```

**Reply by HarelM** [reply]
Yea, this doesn't work for Angular, I couldn't find decent documentation of configuring webpack or babel unfortunately...
It's good that those work for React though.

**Reply by msbarry** [reply]
Is this an angular app directly importing maplibre-gl? Or an angular app importing a library that imports maplibre-gl? If it's the latter you may need a `!` on both the maplibre-gl and library imports.

**Reply by HarelM** [reply]
I'm trying to make the showcase (demo site) of ngx maplibre to work. The "!" doesn't work, it gives a compilation error.

**Reply by birkskyum** [reply]
This could be a way forward. 
- #3228

**Reply by HarelM** [reply]
That's also a possible solution, maybe we can also later on generate cjs and mjs files so that these can be consumed properly and maybe solve this issue that way - keep a commonjs file with "old" syntax and continue forward with the new syntax in a different file, maybe use a major version change to do the flip, IDK, this entire experience reminded me of the old dll hell with c-c++... Just horrible... 😕

**Reply by birkskyum** [reply]
I think the value proposition of compiling async/away to promises is quite good.
- We get to use it now to refactor the code, and can start replacing callbacks
- We don't break anyones code (except the ones hit already, which can be resolved in a new patch)
- It adds just 1kb to the bundle, and that's even a temporary measure until we can safely ship async/await (it clearly didn't go as smooth as we hoped)

**Reply by HarelM** [reply]
Don't get me wrong, I think it's a good solution, I'm just bummed with the current state of web development when we are talking about features that are not that new...

**Reply by msbarry** [reply]
Transpiling now then breaking out a "modern" more compact mjs build from a larger "legacy" cjs build in a future release sounds like a good path forward!

**Comment by birkskyum** [comment, marked as answer]
This will be resolved in:
- https://github.com/maplibre/maplibre-gl-js/pull/3229

**Reply by birkskyum** [reply]
If anyone can confirm the patch, that would be great.

**Reply by mircowidmer** [reply]
Release 3.5.1 works perfectly, thank you very much! 🥇 👍

**Reply by adevinwild** [reply]
Amazing work @birkskyum 🎉
